Sources/AWSLambdaRuntimeCore/Lambda.swift

```diff
@@ -130,6 +135,14 @@ public enum Lambda {
                 }
             }
 
+            // TODO: this is broken, the tracer tries to eat cake and have cake: flash on eventLoop then syncShutdown
+            // fix in XRayRecorder
+            tracer.shutdown { error in
```

Comment:
maybe move this to L129

Reply:
@tomerd 

the problem is that the current implementation in `tracer.shutdown`:
- tries to flush first (on created or shared `EventLoop`),
- waits for it to finish,
- then shutdowns UDP client (which shares or owns the `EventLoop`).

this does not work well if lifecycle and tracer share the `EventLoop`:

```
Precondition failed: BUG DETECTED: wait() must not be called when on an EventLoop.
```

Not sure what the best solution is, I am thinking to remove flushing operation from shutdown,
that way client could always flush on the loop and then shutdown would just close all resources without trying to make any operation on the `EventLoop`.

@ktoso [`TracingInstrument`](https://github.com/slashmo/gsoc-swift-tracing/blob/fe80d764ad225b1dfd06dcb57d08b5e3485662f9/Sources/TracingInstrumentation/TracingInstrument.swift#L19) defined in `swift-tracing` only defines interface to sync flush.

```swift
public protocol TracingInstrument: Instrument {
  func forceFlush()
  // ...
}
```

for reference types in [`AWSXRayRecorder`](https://github.com/pokryfka/aws-xray-sdk-swift/blob/c505ac2d75ce838311365a40a7725cdfdfdcd2ec/Sources/AWSXRayRecorder/Recorder%2BNIO.swift#L17)

```swift
class XRayRecorder {
  // tries to flush firs
  public func shutdown(_ callback: ((Error?) -> Void)? = nil) { }
 // sync flush 
  public func wait(_ callback: ((Error?) -> Void)? = nil)
}

extension XRayRecorder {
    public func flush(on eventLoop: EventLoop) -> EventLoopFuture<Void> {}
}
```

Reply:
This is somewhat of a layering question.

swift-tracing cannot know about nio, so nothing about ELs or similar. 

We _could_ expose a `shutdown(callback)` if needed; question being, should it also flush, does flush need a callback as well then? Flushing is best-effort today and simply a signal to the tracer to try to flush.

The tracing instruments do not define shutdown because, similar like metrics, it is kind of assumed you're managing its lifecycle. We thought that normally you'd likely hook into `swift-lifecycle` with your tracer, and that'd be a specifc type there so you can do whatever you want.

Open to ideas here, what should we co in the API layer to help?
Would it help if we offered hooks that can interop with swift-lifecycle style callbacks nicely?

```swift
// pseudo code
let migrator = DatabaseMigrator()
lifecycle.register(
    label: "tracing",
    start: .sync { bootstrap tracer},
    shutdown: .async(shutdown tracer)
)
```

```swift
// this is again one of those examples which highlight the need for:
https://github.com/swift-server/swift-service-lifecycle/issues/11
// because we could express it as:
lifecycle.register(
    label: "tracing",
    start: .sync { 
         let tracer = MyTracer()
         Instrumentation.bootstrap(...)
         return tracer
    },
    shutdown: .async { tracer in tracer.shutdown(callback: $)) }
)
```

?

Reply:
@ktoso 

> OR, do we need a shutdown(callback) on all instruments? 

this is related but could be handled outside of `TracingInstrument` API IMO

my problem is more with flushing:
- in my case (but I expect this is going to be common) I emit segments (spans) data using UDP client which uses SwiftNIO `EventLoop`
- I would like to share the `EventLoop` if possible, this is already implemented and works
- in case of lambda I need to flush after each invvocation
- `XRayRecoder` provides a method to flush on provided loop which allows to "async flush" but them hop in within invocation
- now, the problem is, `TracingInstrument` only defined sync `forceFlush` - which I provided as ~`flush(eventLoop).wait()`; this will work but not if EventLoop is shared

TL;DR I want to change API to `TracingInstrument` as soon as its release.
I cannot use its `forceFlush` if XRaySDK shares event loop with lambda runtime (which it should...)

Reply:
I'd argue forceFlush() is a _signal_ not a blocking function. don't wait on it, do your best, and your shutdown _can_ wait until flushes are complete. Wouldn't this solve the issue?